### PR TITLE
Moves TestNativeHistogram_MulDivOperator to promql testing framework

### PR DESCRIPTION
**Pull request by NeerajGartia21**
related to: #13563 
Also this PR adds support of Inf and NaN as numbers in histogram buckets
@beorn7 @fatsheep9146 please review !

### CODE REVIEW

File: promql/promqltest/testdata/native_histograms.test [hunk 1]
```diff
@@ -718,6 +718,48 @@ eval instant at 10m histogram_fraction(-Inf, +Inf, histogram_fraction_4)
 eval instant at 10m histogram_sum(scalar(histogram_fraction(-Inf, +Inf, sum(histogram_fraction_4))) * histogram_fraction_4)
     {} 100
 
+# Apply multiplication and division operator to histogram
```
Comment (beorn7) [inline promql/promqltest/testdata/native_histograms.test:721]: Nit: End sentence with period.

File: promql/parser/lex.go [hunk 2]
```diff
@@ -617,6 +617,16 @@ func lexBuckets(l *Lexer) stateFn {
 		l.bracketOpen = false
 		l.emit(RIGHT_BRACKET)
 		return lexHistogram
+	case isAlpha(r):
```
Comment (beorn7) [inline promql/parser/lex.go:620]: Could you add a test case for this to lex_test.go?

**Review by beorn7 — CHANGES_REQUESTED** [review]
Thank you very much. Looks good in general.

I'll check out the tests in detail once #14585 is merged and we can actually run the tests.

In the meantime:
- Let's add a test case for the changed lexer (there are many cases already in lex_test.go, should be very easy to add a buckets with NaN and Inf).
- Please also remove the converted test `TestNativeHistogram_MulDivOperator`.

File: promql/parser/lex.go (hunk 2 above)
Comment (NeerajGartia21) [reply promql/parser/lex.go:620]: added!

File: promql/promqltest/testdata/native_histograms.test (hunk 1 above)
Comment (NeerajGartia21) [reply promql/promqltest/testdata/native_histograms.test]: done !

**Comment by NeerajGartia21** [comment]
> I'll check out the tests in detail once https://github.com/prometheus/prometheus/pull/14585 is merged and we can actually run the tests.

Cool. I've added the tests for changes in lexer and removed the added tests from `engine.go` file.

**Comment by beorn7** [comment]
#14585 is merged. Please rebase this so that we can see if the tests pass.

**Comment by NeerajGartia21** [comment]
Rebased! And All tests are passed!

File: promql/promqltest/testdata/native_histograms.test [hunk 3]
```diff
@@ -718,6 +718,48 @@ eval instant at 10m histogram_fraction(-Inf, +Inf, histogram_fraction_4)
 eval instant at 10m histogram_sum(scalar(histogram_fraction(-Inf, +Inf, sum(histogram_fraction_4))) * histogram_fraction_4)
     {} 100
 
+# Apply multiplication and division operator to histogram.
+load 10m
+    histogram_mul_div {{schema:0 count:21 sum:33 z_bucket:3 z_bucket_w:0.001 buckets:[3 3 3] n_buckets:[3 3 3]}}x1
+	float_series_3 3+0x1
+	float_series_0 0+0x1
```
Comment (beorn7) [inline promql/promqltest/testdata/native_histograms.test:725]: Some lines use blanks and some lines use tabs.

At least within the same test, we should be consistent. I recommend to use just blanks everywhere (your editor probably has an "untabify" function – it'l literally called `untabify` in emacs, for example).

File: promql/promqltest/testdata/native_histograms.test [hunk 4]
```diff
@@ -718,6 +718,48 @@ eval instant at 10m histogram_fraction(-Inf, +Inf, histogram_fraction_4)
...
+    {} {{schema:0 count:0 sum:0 z_bucket:0 z_bucket_w:0.001 buckets:[0 0 0] n_buckets:[0 0 0]}}
+
+eval instant at 10m float_series_0*histogram_mul_div
+    {} {{schema:0 count:0 sum:0 z_bucket:0 z_bucket_w:0.001 buckets:[0 0 0] n_buckets:[0 0 0]}}
+
+eval instant at 10m histogram_mul_div/0
+    {} {{schema:0 count:Inf sum:Inf z_bucket:Inf z_bucket_w:0.001 buckets:[Inf Inf Inf] n_buckets:[Inf Inf Inf]}}
+
+eval instant at 10m histogram_mul_div/float_series_0
+    {} {{schema:0 count:Inf sum:Inf z_bucket:Inf z_bucket_w:0.001 buckets:[Inf Inf Inf] n_buckets:[Inf Inf Inf]}}
```
Comment (beorn7) [inline promql/promqltest/testdata/native_histograms.test:761]: Let's add a TODO here that we plan to change this behavior (linking to #13934).

**Review by beorn7 — CHANGES_REQUESTED** [review]
Looks great. Just a tab vs. spaces nit and a request for a comment.

And one more thing: Maybe add a new test (now that it is so easy) evaluating `histogram_mul_div*0/0` (to see `NaN`s popping up).

**Comment by beorn7** [comment]
Also, we have a merge conflict now…

File: promql/promqltest/testdata/native_histograms.test (hunk 3 above)
Comment (NeerajGartia21) [reply promql/promqltest/testdata/native_histograms.test]: done!

File: promql/promqltest/testdata/native_histograms.test (hunk 4 above)
Comment (NeerajGartia21) [reply promql/promqltest/testdata/native_histograms.test:762]: Done!

**Comment by NeerajGartia21** [comment]
Resolved all the merge conflicts and applied all the suggested changes!

**Pull request merged**